[WIP :construction: ] Automatic qubit management (#1788)
PennyLaneAI/catalyst: doc/releases/changelog-dev.md

  Array([0.25, 0.25, 0.25, 0.25], dtype=float64))
  ```

* Catalyst now supports automatic qubit management.
  [(#1788)](https://github.com/PennyLaneAI/catalyst/pull/1788)

  The number of wires does not need to be speficied during device initialization,

>   and can be automatically allocated whenever a new wire is encountered.

Reviewer: Let's try not to be too descriptive about _how_ it works, since the implementation could change. We can focus on what changes for the user.
```suggestion
  and instead will be automatically managed by the Catalyst Runtime.
```

  ```python
  @qjit
  def workflow():
      dev = qml.device("lightning.qubit") # no wires here!
      @qml.qnode(dev)
      def circuit():
          qml.PauliX(wires=2)
          return qml.probs()
      return circuit()

  print(workflow())
  ```

  ```pycon
  [0. 1. 0. 0. 0. 0. 0. 0.]
  ```

  In this example, the number of wires is not specified at device initialization.

>   When we encounter an X gate on `wires=2`, catalyst automatically allocates 3 wires, and applies

Reviewer: ```suggestion
  When we encounter an X gate on `wires=2`, catalyst automatically expands the size of the qubit
  register to include the requested wire index. Here, the register will contain at least 3 qubits
  after the X operation.
```
I'm using at least here in case we want to use more efficient allocation schemes. For instance we could always start with a minimum, or increase by 2, or whatever other scheme we might have.

Author: https://github.com/PennyLaneAI/catalyst/pull/1788/commits/ce8e78e609d5bd3ec83b89597a2007132784a77a all changelog suggestions applied here

>   an X gate to the wire with index `2`. The result is the state |001>, giving the probabilities
>   above.

Reviewer: ```suggestion
  As a result, we can see the QNode returning the probabilities for the state |001>, meaning 3 wires
  were allocated in total.
```

>   This feature can be turned on by simply not supplying a `wires` argument to the device.

Reviewer: ```suggestion
  This feature can be turned on by omitting the `wires` argument to the device.
```
